Number_types/include/CGAL/cpp_float.h

```diff
@@ -0,0 +1,487 @@
+#endif
+  }
+
+} // namespace internal
+
+#if 0 // needs C++20
+  template <class T>
+  void fmt(const T& t)
+  {
+    std::cout << std::format("{:b}", t) << std::endl;
+  }
+#endif
+
+class cpp_float {
+#ifdef CGAL_CPPF
+  boost::multiprecision::cpp_rational rat;
+#endif
+  boost::multiprecision::cpp_int man;
+  int exp; /* The number man (an integer) * 2 ^ exp  */
+
+public:
+  cpp_float()
+    :
+#ifdef CGAL_CPPF
+    rat(),
+#endif
+    man(), exp()
+  {}
+
+  cpp_float(short i)
+    :
+#ifdef CGAL_CPPF
+    rat(i),
+#endif
+    man(i),exp(0)
+  {}
+
+  cpp_float(int i)
+    :
+#ifdef CGAL_CPPF
+    rat(i),
+#endif
+    man(i),exp(0)
+  {}
+
+  cpp_float(long i)
+    :
+#ifdef CGAL_CPPF
+    rat(i),
+#endif
+    man(i),exp(0)
+  {}
+#ifdef CGAL_CPPF
+  cpp_float(const boost::multiprecision::cpp_int& man,  int exp, const boost::multiprecision::cpp_rational& rat)
+    : rat(rat), man(man),exp(exp)
+  {}
+#else
+
+  cpp_float(const boost::multiprecision::cpp_int& man, int exp)
+      : man(man), exp(exp)
+  {}
+#endif
+  cpp_float(double d)
+
+#ifdef CGAL_CPPF
+   : rat(d)
+#endif
+  {
+    //std::cout << "\ndouble = " << d << std::endl;
+    using boost::uint64_t;
+    union {
+#ifdef CGAL_LITTLE_ENDIAN
+      struct { uint64_t man:52; uint64_t exp:11; uint64_t sig:1; } s;
+#else /* CGAL_BIG_ENDIAN */
+      //WARNING: untested!
+      struct { uint64_t sig:1; uint64_t exp:11; uint64_t man:52; } s;
+#endif
+      double d;
+    } u;
+    u.d = d;
+
+    uint64_t m;
+    uint64_t dexp = u.s.exp;
+    CGAL_assertion_msg(dexp != 2047, "Creating an cpp_float from infinity or NaN.");
+    if (dexp == 0) {
+      if (d == 0) { exp=0; return; }
+      else { // denormal number
+        m = u.s.man;
+        ++dexp;
+      }
+    } else {
+      m = (1LL << 52) |  u.s.man;
+    }
+
+
+    int idexp = (int)dexp;
+    idexp -= 1023;
+
+    // std::cout << "m    = "  << m << std::endl;
+    // std::cout << "idexp = " << idexp << std::endl;
+
+    int shifted = internal::low_bit(m);
+
+    m >>= shifted;
+
+    int nbits = internal::high_bit(m);
+    // std::cout << "nbits = " << nbits << std::endl;
+
+    exp = idexp - nbits;
+    if(u.s.sig){
+      m = -m;
+    }
+    // std::cout << "m = " << m << " * 2^" << exp  << std::endl;
+    // fmt(m);
+    man = boost::multiprecision::cpp_int(m);
+  }
+
+  friend std::ostream& operator<<(std::ostream& os, const cpp_float& m)
+  {
+    return os << m.man << " * 2 ^ " << m.exp << " ( " << m.to_double() << ") ";
+  }
+
+  friend cpp_float operator-(cpp_float const&x)
+  {
+#ifdef CGAL_CPPF
+    return cpp_float(-x.man,x.exp, -x.rat);
+#else
+    return cpp_float(-x.man,x.exp);
+#endif
+  }
+
+  cpp_float& operator*=(const cpp_float& other)
+  {
+#ifdef CGAL_CPPF
+    rat *= other.rat;
+#endif
+    man *= other.man;
+    exp += other.exp;
+    return *this;
+  }
+
+
+  friend
+  cpp_float operator*(const cpp_float& a, const cpp_float&b){
+    return cpp_float(a.man*b.man, a.exp+b.exp);
```

Comment:
It could be useful to have a constructor for cpp_float that accepts an expression as first argument so the product can be evaluated directly in the new cpp_float instead of first in a cpp_int and then copied. Or at least one that accepts a cpp_int rvalue so it can be moved instead of copied.

Reply:
It in fact IS useful.